oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/dao/BanyanDBTraceQueryDAO.java

```diff
@@ -0,0 +1,156 @@
+        } else {
+            query = new TraceQuery(BanyanDBSchema.NAME, BASIC_QUERY_PROJ);
+        }
+        if (minDuration != 0) {
+            // duration >= minDuration
+            query.appendCondition(PairQueryCondition.LongQueryCondition.ge("duration", minDuration));
+        }
+        if (maxDuration != 0) {
+            // duration <= maxDuration
+            query.appendCondition(PairQueryCondition.LongQueryCondition.le("duration", maxDuration));
+        }
+
+        if (!Strings.isNullOrEmpty(serviceId)) {
+            query.appendCondition(PairQueryCondition.StringQueryCondition.eq("service_id", serviceId));
+        }
+
+        if (!Strings.isNullOrEmpty(serviceInstanceId)) {
+            query.appendCondition(PairQueryCondition.StringQueryCondition.eq("service_instance_id", serviceInstanceId));
+        }
+
+        if (!Strings.isNullOrEmpty(endpointId)) {
+            query.appendCondition(PairQueryCondition.StringQueryCondition.eq("endpoint_id", endpointId));
+        }
+
+        switch (traceState) {
+            case ERROR:
+                query.appendCondition(PairQueryCondition.LongQueryCondition.eq("state", (long) BanyanDBSchema.TraceState.ERROR.getState()));
+                break;
+            case SUCCESS:
+                query.appendCondition(PairQueryCondition.LongQueryCondition.eq("state", (long) BanyanDBSchema.TraceState.SUCCESS.getState()));
+                break;
+            default:
+                query.appendCondition(PairQueryCondition.LongQueryCondition.eq("state", (long) BanyanDBSchema.TraceState.ALL.getState()));
+                break;
+        }
+
+        switch (queryOrder) {
+            case BY_START_TIME:
+                query.setOrderBy(new TraceQuery.OrderBy("start_time", TraceQuery.OrderBy.Type.DESC));
+                break;
+            case BY_DURATION:
+                query.setOrderBy(new TraceQuery.OrderBy("duration", TraceQuery.OrderBy.Type.DESC));
+                break;
+        }
+
+        query.setLimit(limit);
+        query.setOffset(from);
+
+        // build request
+        TraceQueryResponse response = this.getClient().query(query);
+        TraceBrief brief = new TraceBrief();
+        brief.setTotal(response.size());
+        brief.getTraces().addAll(response.getEntities().stream().map(entity -> {
+            BasicTrace trace = new BasicTrace();
+            trace.setDuration(((Long) entity.getFields().get(3).getValue()).intValue());
+            trace.setStart(String.valueOf(entity.getFields().get(4).getValue()));
+            trace.setSegmentId(entity.getId());
+            trace.setError(((Long) entity.getFields().get(1).getValue()).intValue() == 1);
+            trace.getTraceIds().add((String) entity.getFields().get(0).getValue());
+            trace.getEndpointNames().add(IDManager.EndpointID.analysisId(
+                    (String) entity.getFields().get(2).getValue()
+            ).getEndpointName());
+            return trace;
```

Comment:
Please make sure the codes polished. When you face a 0-N field indices, you should make sure they are ordered.

Reply:
And where is the request fields? I am not sure how BanyanDB server implements the logic now when they are loss, but we access them through indices, they must be declared. If not, BanyanDB should response nothing.

Reply:
> Please make sure the codes polished. When you face a 0-N field indices, you should make sure they are ordered.

Sorted.

Reply:
> And where is the request fields? I am not sure how BanyanDB server implements the logic now when they are loss, but we access them through indices, they must be declared. If not, BanyanDB should response nothing.

What do you mean by request fields here?